### PR TITLE
feat(generator): support explicit routing headers

### LINKED ISSUE #9121: Support `routing_parameters` in the generator
At the moment, these are only used in the Bigtable Data API (which we handwrite) and Cloud Run (which we do not generate). Here is an example:
https://github.com/googleapis/googleapis/blob/36a7e1dd18aecda768b560a7e5c155f79c867370/google/bigtable/v2/bigtable.proto#L65-L73

For how we should process these fields, see the `google.api.routing` proto: https://github.com/googleapis/googleapis/blob/177c62af7f181a42b91def98dec2e46efdb07b5e/google/api/routing.proto

### PR DESCRIPTION
**Pull request by dbolduc**
Fixes #9121 

The first and last commits are interesting. The middle two are not.

We do not pass our variables through `service_method_vars`. It is easier to work with the data directly.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9368)

### CODE REVIEW

File: generator/internal/metadata_decorator_generator.cc [hunk 1]
```diff
@@ -58,12 +58,13 @@ std::string SetMetadataText(google::protobuf::MethodDescriptor const& method,
     if (std::all_of(
             kv.second.begin(), kv.second.end(),
             [](RoutingParameter const& rp) { return rp.pattern == "(.*)"; })) {
-      for (auto i = 0; i != kv.second.size(); ++i){
-        auto const& field = kv.second[i].field_name;
-        text += i == 0 ? "  " : " else ";
-        text += "if (!request." + field + "().empty()) {\n";
-        text += "    params.push_back(\"" + kv.first + "=\" + request." + field + "());\n";
+      bool initial_if = true;
```
Comment (coryan) [inline generator/internal/metadata_decorator_generator.cc:61]: consider:

```cc
char const* sep = "  ";
for (auto const& rp : kv.second) {
  text += sep;
  ....
  ...
  sep = " else ";
}
```

File: generator/internal/metadata_decorator_generator.cc [hunk 2]
```diff
@@ -32,14 +33,64 @@ std::string SetMetadataText(google::protobuf::MethodDescriptor const& method,
...
+    }
+    // If the method does not have a `google.api.routing` or `google.api.http`
+    // annotation, we do not send the "x-goog-request-params" header.
+    return "  SetMetadata(" + context + ");";
   }
-  // If the method does not have a `google.api.http` annotation, we do not send
-  // the "x-goog-request-params" header.
-  return "  SetMetadata(" + context + ");";
+
+  // clang-format off
```
Comment (coryan) [inline generator/internal/metadata_decorator_generator.cc:50]: Ah, yes, that is the thing turning off formatting.

**Review by coryan — COMMENTED** [review]
Do you want to add tests for the async cases?  Or streaming?

File: generator/internal/metadata_decorator_generator.cc (hunk 1 above)
Comment (dbolduc) [reply generator/internal/metadata_decorator_generator.cc]: Done.

File: generator/internal/metadata_decorator_generator.cc [hunk 3]
```diff
@@ -58,12 +58,13 @@ std::string SetMetadataText(google::protobuf::MethodDescriptor const& method,
     if (std::all_of(
             kv.second.begin(), kv.second.end(),
             [](RoutingParameter const& rp) { return rp.pattern == "(.*)"; })) {
-      for (auto i = 0; i != kv.second.size(); ++i){
-        auto const& field = kv.second[i].field_name;
-        text += i == 0 ? "  " : " else ";
-        text += "if (!request." + field + "().empty()) {\n";
-        text += "    params.push_back(\"" + kv.first + "=\" + request." + field + "());\n";
+      bool initial_if = true;
+      for (auto const& rp : kv.second){
```
[the remark this answers is not in the document]
Comment (dbolduc) [reply generator/internal/metadata_decorator_generator.cc:62]: It was a deliberate choice to keep the lines intact

**Comment by dbolduc** [comment]
> Do you want to add tests for the async cases? Or streaming?

I dropped the routing annotation on a streaming read and unary rpc where we generate async versions of the functions. `IsContextMDValid` revealed that if no routing matches were found, I wasn't sending any metadata at all. Yay for testing? Boo for self esteem.

**Review by coryan — APPROVED** [review]
I think we still need to figure out streaming writes and streaming read-write RPCs, but those are (a) very rare, and (b) already broken I think.

**Pull request merged**